File: test/evm/precompiles/ripemd160_test.zig
const std = @import("std");
const testing = std.testing;
const Evm = @import("evm");
const primitives = @import("primitives");
const Address = primitives.Address.Address;

// Convenience aliases
const ripemd160 = Evm.precompiles.ripemd160;
const precompiles = Evm.precompiles.precompiles;
const PrecompileOutput = Evm.precompiles.precompile_result.PrecompileOutput;
const PrecompileError = Evm.precompiles.precompile_result.PrecompileError;
const ChainRules = Evm.hardforks.chain_rules;

// Known RIPEMD160 test vectors from official specification
// and Bitcoin Core implementation
const EMPTY_HASH = [_]u8{ 0x9c, 0x11, 0x85, 0xa5, 0xc5, 0xe9, 0xfc, 0x54, 0x61, 0x28, 0x08, 0x97, 0x7e, 0xe8, 0xf5, 0x48, 0xb2, 0x25, 0x8d, 0x31 };

const ABC_HASH = [_]u8{ 0x8e, 0xb2, 0x08, 0xf7, 0xe0, 0x5d, 0x98, 0x7a, 0x9b, 0x04, 0x4a, 0x8e, 0x98, 0xc6, 0xb0, 0x87, 0xf1, 0x5a, 0x0b, 0xfc };

const MESSAGE_DIGEST_HASH = [_]u8{ 0x5d, 0x06, 0x89, 0xef, 0x49, 0xd2, 0xfa, 0xe5, 0x72, 0xb8, 0x81, 0xb1, 0x23, 0xa8, 0x5f, 0xfa, 0x21, 0x59, 0x5f, 0x36 };

const ALPHABET_HASH = [_]u8{ 0xf7, 0x1c, 0x27, 0x10, 0x9c, 0x69, 0x2c, 0x1b, 0x56, 0xbb, 0xdc, 0xeb, 0x5b, 0x9d, 0x28, 0x65, 0xb3, 0x70, 0x8d, 0xbc };

const LONG_STRING_HASH = [_]u8{ 0x12, 0xa0, 0x53, 0x38, 0x4a, 0x9c, 0x0c, 0x88, 0xe4, 0x05, 0xa0, 0x6c, 0x27, 0xdc, 0xf4, 0x9a, 0xda, 0x62, 0xeb, 0x2b };

const ALPHANUM_HASH = [_]u8{ 0xb0, 0xe2, 0x0b, 0x6e, 0x31, 0x16, 0x64, 0x02, 0x86, 0xed, 0x3a, 0x87, 0xa5, 0x71, 0x30, 0x79, 0xb2, 0x1f, 0x51, 0x89 };

const REPEATED_DIGITS_HASH = [_]u8{ 0x9b, 0x75, 0x2e, 0x45, 0x57, 0x3d, 0x4b, 0x39, 0xf4, 0xdb, 0xd3, 0x32, 0x3c, 0xab, 0x82, 0xbf, 0x63, 0x32, 0x6b, 0xfb };

const MILLION_A_HASH = [_]u8{ 0x52, 0x78, 0x32, 0x43, 0xc1, 0x69, 0x7b, 0xdb, 0xe1, 0x6d, 0x37, 0xf9, 0x7f, 0x68, 0xf0, 0x83, 0x25, 0xdc, 0x15, 0x28 };

test "RIPEMD160 gas calculation matches Ethereum specification" {
    // Empty input: 600 base gas
    try testing.expectEqual(@as(u64, 600), ripemd160.calculate_gas(0));

    // 1 byte: 600 + 120 * 1 = 720
    try testing.expectEqual(@as(u64, 720), ripemd160.calculate_gas(1));

    // 32 bytes: 600 + 120 * 1 = 720
    try testing.expectEqual(@as(u64, 720), ripemd160.calculate_gas(32));

    // 33 bytes: 600 + 120 * 2 = 840
    try testing.expectEqual(@as(u64, 840), ripemd160.calculate_gas(33));

    // 64 bytes: 600 + 120 * 2 = 840
    try testing.expectEqual(@as(u64, 840), ripemd160.calculate_gas(64));

    // 65 bytes: 600 + 120 * 3 = 960
    try testing.expectEqual(@as(u64, 960), ripemd160.calculate_gas(65));
}

test "RIPEMD160 gas calculation overflow protection" {
    // Should handle normal sizes without error
    try testing.expectEqual(@as(u64, 600), try ripemd160.calculate_gas_checked(0));
    try testing.expectEqual(@as(u64, 720), try ripemd160.calculate_gas_checked(1));

    // Test with large but valid input size
    const large_size = 1024 * 1024; // 1MB
    const expected_gas = 600 + 120 * ((large_size + 31) / 32);
    try testing.expectEqual(expected_gas, try ripemd160.calculate_gas_checked(large_size));
}

test "RIPEMD160 output size is always 32 bytes" {
    try testing.expectEqual(@as(usize, 32), ripemd160.get_output_size(0));
    try testing.expectEqual(@as(usize, 32), ripemd160.get_output_size(1));
    try testing.expectEqual(@as(usize, 32), ripemd160.get_output_size(100));
    try testing.expectEqual(@as(usize, 32), ripemd160.get_output_size(1000));
}

test "RIPEMD160 call validation" {
    // Should succeed with sufficient gas
    try testing.expect(ripemd160.validate_call(0, 600));
    try testing.expect(ripemd160.validate_call(1, 720));
    try testing.expect(ripemd160.validate_call(32, 720));

    // Should fail with insufficient gas
    try testing.expect(!ripemd160.validate_call(0, 599));
    try testing.expect(!ripemd160.validate_call(1, 719));
    try testing.expect(!ripemd160.validate_call(33, 839));
}

test "ripemd160 execute empty input" {
    var output_buffer: [32]u8 = undefined;
    const result = ripemd160.execute(&[_]u8{}, &output_buffer, 700);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 600), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify hash matches known empty string hash
    // First 12 bytes should be zeros (padding)
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    // Next 20 bytes should be the hash
    try testing.expectEqualSlices(u8, &EMPTY_HASH, output_buffer[12..32]);
}

test "ripemd160 execute 'abc'" {
    var output_buffer: [32]u8 = undefined;
    const result = ripemd160.execute("abc", &output_buffer, 800);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 720), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &ABC_HASH, output_buffer[12..32]);
}

test "ripemd160 execute 'message digest'" {
    var output_buffer: [32]u8 = undefined;
    const result = ripemd160.execute("message digest", &output_buffer, 800);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 720), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &MESSAGE_DIGEST_HASH, output_buffer[12..32]);
}

test "ripemd160 execute alphabet" {
    var output_buffer: [32]u8 = undefined;
    const input = "abcdefghijklmnopqrstuvwxyz";
    const result = ripemd160.execute(input, &output_buffer, 800);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 720), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &ALPHABET_HASH, output_buffer[12..32]);
}

test "ripemd160 execute long string" {
    var output_buffer: [32]u8 = undefined;
    const input = "abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq";
    const result = ripemd160.execute(input, &output_buffer, 1000);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 840), result.get_gas_used()); // 2 words
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &LONG_STRING_HASH, output_buffer[12..32]);
}

test "ripemd160 execute alphanumeric" {
    var output_buffer: [32]u8 = undefined;
    const input = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
    const result = ripemd160.execute(input, &output_buffer, 1000);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 840), result.get_gas_used()); // 2 words
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &ALPHANUM_HASH, output_buffer[12..32]);
}

test "ripemd160 execute repeated digits" {
    var output_buffer: [32]u8 = undefined;
    const input = "12345678901234567890123456789012345678901234567890123456789012345678901234567890";
    const result = ripemd160.execute(input, &output_buffer, 1000);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 960), result.get_gas_used()); // 3 words
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &REPEATED_DIGITS_HASH, output_buffer[12..32]);
}

test "ripemd160 execute out of gas" {
    const input_data = "abc";
    var output_buffer: [32]u8 = undefined;

    // Provide insufficient gas (need 720, only provide 719)
    const result = ripemd160.execute(input_data, &output_buffer, 719);

    try testing.expect(result.is_failure());
    try testing.expectEqual(PrecompileError.OutOfGas, result.get_error().?);
    try testing.expectEqual(@as(u64, 0), result.get_gas_used());
}

test "ripemd160 execute insufficient output buffer" {
    const input_data = "abc";
    var output_buffer: [31]u8 = undefined; // Too small for 32-byte output

    const result = ripemd160.execute(input_data, &output_buffer, 800);

    try testing.expect(result.is_failure());
    try testing.expectEqual(PrecompileError.ExecutionFailed, result.get_error().?);
}

test "precompile dispatcher ripemd160 integration" {
    const ripemd160_address: Address = [_]u8{ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x03 };

    // Check address detection
    try testing.expect(precompiles.is_precompile(ripemd160_address));

    // Check availability with default chain rules
    const chain_rules = ChainRules.DEFAULT;
    try testing.expect(precompiles.is_available(ripemd160_address, chain_rules));
}

test "precompile dispatcher execute ripemd160" {
    const ripemd160_address: Address = [_]u8{ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0x03 };

    var output_buffer: [32]u8 = undefined;

    const chain_rules = ChainRules.DEFAULT;
    const result = precompiles.execute_precompile(ripemd160_address, "abc", &output_buffer, 800, chain_rules);

    try testing.expect(result.is_success());
    try testing.expectEqual(@as(u64, 720), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &ABC_HASH, output_buffer[12..32]);
}

test "ripemd160 large input stress test" {
    const allocator = testing.allocator;

    // Test with 1KB of data
    const input_size = 1024;
    const input_data = try allocator.alloc(u8, input_size);
    defer allocator.free(input_data);

    var output_buffer: [32]u8 = undefined;

    // Fill with test pattern
    for (input_data, 0..) |*byte, i| {
        byte.* = @intCast(i % 256);
    }

    const result = ripemd160.execute(input_data, &output_buffer, 5000);
    try testing.expect(result.is_success());
    // 1024 bytes = 32 words, so 600 + 120*32 = 4440 gas
    try testing.expectEqual(@as(u64, 4440), result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());
}

test "ripemd160 million 'a' test" {
    const allocator = testing.allocator;

    // This is a standard test vector but requires significant memory
    // Only run if we have enough memory available
    const million = 1_000_000;
    const input_data = allocator.alloc(u8, million) catch |err| {
        if (err == error.OutOfMemory) {
            // Skip test if we can't allocate
            return;
        }
        return err;
    };
    defer allocator.free(input_data);

    // Fill with 'a'
    @memset(input_data, 'a');

    var output_buffer: [32]u8 = undefined;

    // Calculate required gas: 600 + 120 * ceil(1000000/32) = 600 + 120 * 31250 = 3,750,600
    const required_gas = ripemd160.calculate_gas(million);
    const result = ripemd160.execute(input_data, &output_buffer, required_gas);

    try testing.expect(result.is_success());
    try testing.expectEqual(required_gas, result.get_gas_used());
    try testing.expectEqual(@as(usize, 32), result.get_output_size());

    // Verify output format and hash
    try testing.expectEqualSlices(u8, &[_]u8{0} ** 12, output_buffer[0..12]);
    try testing.expectEqualSlices(u8, &MILLION_A_HASH, output_buffer[12..32]);
}

File: src/crypto/ripemd160.test.ts
/**
 * RIPEMD160 Tests
 *
 * Test vectors from official RIPEMD160 specification
 * and Bitcoin Core implementation.
 */

import { describe, it, expect } from 'vitest';
import { Ripemd160 } from './ripemd160.js';

describe('Ripemd160', () => {
  describe('hash', () => {
    it('empty string', () => {
      const expected = new Uint8Array([
        0x9c, 0x11, 0x85, 0xa5, 0xc5, 0xe9, 0xfc, 0x54,
        0x61, 0x28, 0x08, 0x97, 0x7e, 0xe8, 0xf5, 0x48,
        0xb2, 0x25, 0x8d, 0x31,
      ]);
      const result = Ripemd160.hashString('');
      expect(result).toEqual(expected);
    });

    it('single byte "a"', () => {
      const expected = new Uint8Array([
        0x0b, 0xdc, 0x9d, 0x2d, 0x25, 0x6b, 0x3e, 0xe9,
        0xda, 0xae, 0x34, 0x7b, 0xe6, 0xf4, 0xdc, 0x83,
        0x5a, 0x46, 0x7f, 0xfe,
      ]);
      const result = Ripemd160.hashString('a');
      expect(result).toEqual(expected);
    });

    it('"abc"', () => {
      const expected = new Uint8Array([
        0x8e, 0xb2, 0x08, 0xf7, 0xe0, 0x5d, 0x98, 0x7a,
        0x9b, 0x04, 0x4a, 0x8e, 0x98, 0xc6, 0xb0, 0x87,
        0xf1, 0x5a, 0x0b, 0xfc,
      ]);
      const result = Ripemd160.hashString('abc');
      expect(result).toEqual(expected);
    });

    it('"message digest"', () => {
      const expected = new Uint8Array([
        0x5d, 0x06, 0x89, 0xef, 0x49, 0xd2, 0xfa, 0xe5,
        0x72, 0xb8, 0x81, 0xb1, 0x23, 0xa8, 0x5f, 0xfa,
        0x21, 0x59, 0x5f, 0x36,
      ]);
      const result = Ripemd160.hashString('message digest');
      expect(result).toEqual(expected);
    });

    it('lowercase alphabet', () => {
      const expected = new Uint8Array([
        0xf7, 0x1c, 0x27, 0x10, 0x9c, 0x69, 0x2c, 0x1b,
        0x56, 0xbb, 0xdc, 0xeb, 0x5b, 0x9d, 0x28, 0x65,
        0xb3, 0x70, 0x8d, 0xbc,
      ]);
      const result = Ripemd160.hashString('abcdefghijklmnopqrstuvwxyz');
      expect(result).toEqual(expected);
    });

    it('alphanumeric', () => {
      const expected = new Uint8Array([
        0x12, 0xa0, 0x53, 0x38, 0x4a, 0x9c, 0x0c, 0x88,
        0xe4, 0x05, 0xa0, 0x6c, 0x27, 0xdc, 0xf4, 0x9a,
        0xda, 0x62, 0xeb, 0x2b,
      ]);
      const result = Ripemd160.hashString('abcdbcdecdefdefgefghfghighijhijkijkljklmklmnlmnomnopnopq');
      expect(result).toEqual(expected);
    });

    it('mixed case alphanumeric', () => {
      const expected = new Uint8Array([
        0xb0, 0xe2, 0x0b, 0x6e, 0x31, 0x16, 0x64, 0x02,
        0x86, 0xed, 0x3a, 0x87, 0xa5, 0x71, 0x30, 0x79,
        0xb2, 0x1f, 0x51, 0x89,
      ]);
      const result = Ripemd160.hashString('ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789');
      expect(result).toEqual(expected);
    });

    it('repeated digits', () => {
      const expected = new Uint8Array([
        0x9b, 0x75, 0x2e, 0x45, 0x57, 0x3d, 0x4b, 0x39,
        0xf4, 0xdb, 0xd3, 0x32, 0x3c, 0xab, 0x82, 0xbf,
        0x63, 0x32, 0x6b, 0xfb,
      ]);
      const result = Ripemd160.hashString('12345678901234567890123456789012345678901234567890123456789012345678901234567890');
      expect(result).toEqual(expected);
    });

    it('hello', () => {
      const expected = new Uint8Array([
        0x10, 0x8f, 0x07, 0xb8, 0x38, 0x24, 0x12, 0x61,
        0x2c, 0x04, 0x8d, 0x07, 0xd1, 0x3f, 0x81, 0x41,
        0x18, 0x44, 0x5a, 0xcd,
      ]);
      const result = Ripemd160.hashString('hello');
      expect(result).toEqual(expected);
    });

    it('The quick brown fox', () => {
      const expected = new Uint8Array([
        0x37, 0xf3, 0x32, 0xf6, 0x8d, 0xb7, 0x7b, 0xd9,
        0xd7, 0xed, 0xd4, 0x96, 0x95, 0x71, 0xad, 0x67,
        0x1c, 0xf9, 0xdd, 0x3b,
      ]);
      const result = Ripemd160.hashString('The quick brown fox jumps over the lazy dog');
      expect(result).toEqual(expected);
    });
  });

  describe('hash with Uint8Array input', () => {
    it('empty bytes', () => {
      const expected = new Uint8Array([
        0x9c, 0x11, 0x85, 0xa5, 0xc5, 0xe9, 0xfc, 0x54,
        0x61, 0x28, 0x08, 0x97, 0x7e, 0xe8, 0xf5, 0x48,
        0xb2, 0x25, 0x8d, 0x31,
      ]);
      const result = Ripemd160.hash(new Uint8Array([]));
      expect(result).toEqual(expected);
    });

    it('single byte 0x00', () => {
      const expected = new Uint8Array([
        0xc8, 0x1b, 0x94, 0x93, 0x34, 0x20, 0x22, 0x1a,
        0x7a, 0xc0, 0x04, 0xa9, 0x02, 0x42, 0xd8, 0xb1,
        0xd3, 0xe5, 0x07, 0x0d,
      ]);
      const result = Ripemd160.hash(new Uint8Array([0x00]));
      expect(result).toEqual(expected);
    });

    it('single byte 0xFF', () => {
      const expected = new Uint8Array([
        0x2c, 0x0c, 0x45, 0xd3, 0xec, 0xab, 0x80, 0xfe,
        0x06, 0x0e, 0x5f, 0x1d, 0x70, 0x57, 0xcd, 0x2f,
        0x8d, 0xe5, 0xe5, 0x57,
      ]);
      const result = Ripemd160.hash(new Uint8Array([0xFF]));
      expect(result).toEqual(expected);
    });

    it('binary data with all byte values', () => {
      const input = new Uint8Array(256);
      for (let i = 0; i < 256; i++) {
        input[i] = i;
      }
      const result = Ripemd160.hash(input);

      // Result should be 20 bytes
      expect(result.length).toBe(20);

      // Result should be non-zero
      const allZero = result.every(byte => byte === 0);
      expect(allZero).toBe(false);

      // Deterministic
      const result2 = Ripemd160.hash(input);
      expect(result).toEqual(result2);
    });
  });

  describe('edge cases', () => {
    it('exactly 64 bytes', () => {
      const expected = new Uint8Array([
        0x9d, 0xfb, 0x7d, 0x37, 0x4a, 0xd9, 0x24, 0xf3,
        0xf8, 0x8d, 0xe9, 0x62, 0x91, 0xc3, 0x3e, 0x9a,
        0xbe, 0xd5, 0x3e, 0x32,
      ]);
      const result = Ripemd160.hashString('a'.repeat(64));
      expect(result).toEqual(expected);
    });

    it('exactly 128 bytes', () => {
      const expected = new Uint8Array([
        0x8d, 0xfd, 0xfb, 0x32, 0xb2, 0xed, 0x5c, 0xb4,
        0x1a, 0x73, 0x47, 0x8b, 0x4f, 0xd6, 0x0c, 0xc5,
        0xb4, 0x64, 0x8b, 0x15,
      ]);
      const result = Ripemd160.hashString('a'.repeat(128));
      expect(result).toEqual(expected);
    });

    it('large input', () => {
      const input = 'a'.repeat(10000);
      const result = Ripemd160.hash(input);

      // Should be 20 bytes
      expect(result.length).toBe(20);

      // Should be deterministic
      const result2 = Ripemd160.hash(input);
      expect(result).toEqual(result2);
    });
  });

  describe('Bitcoin compatibility', () => {
    it('Bitcoin address generation step', () => {
      // In Bitcoin, addresses use RIPEMD160(SHA256(pubkey))
      // This tests the RIPEMD160 part with a SHA256 hash as input

      // Example SHA256 hash (32 bytes)
      const sha256Hash = new Uint8Array([
        0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07,
        0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
        0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17,
        0x18, 0x19, 0x1a, 0x1b, 0x1c, 0x1d, 0x1e, 0x1f,
      ]);

      const result = Ripemd160.hash(sha256Hash);

      // Result should be 20 bytes (standard Bitcoin pubkey hash size)
      expect(result.length).toBe(20);
    });
  });

  describe('avalanche effect', () => {
    it('small input change causes large hash change', () => {
      const hash1 = Ripemd160.hashString('test message');
      const hash2 = Ripemd160.hashString('test messag3'); // One char different

      // Count differing bytes
      let differences = 0;
      for (let i = 0; i < hash1.length; i++) {
        if (hash1[i] !== hash2[i]) differences++;
      }

      // Expect significant difference (at least 50% of bytes)
      expect(differences).toBeGreaterThanOrEqual(10);
    });
  });
});
